### PR TITLE
man-db: fixes for Linuxbrew

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

File: Formula/man-db.rb [hunk 1]
```diff
@@ -20,6 +25,7 @@ class ManDb < Formula
 
   def install
     resource("libpipeline").stage do
+      ENV.append_to_cflags "-fPIC" unless OS.mac?
```
Comment (sjackman) [inline Formula/man-db.rb]: Why is `-fPIC` needed when `--enable-static --disable-shared` is used below? What's the error message without this line?

File: Formula/man-db.rb [hunk 2]
```diff
@@ -55,7 +61,7 @@ def install
     # https://git.savannah.gnu.org/cgit/man-db.git/commit/?id=056e8c7c012b00261133259d6438ff8303a8c36c
     ENV.append_to_cflags "-Wl,-flat_namespace,-undefined,suppress"
 
-    system "make", "CFLAGS=#{ENV.cflags}"
+    system "make", *("CFLAGS=#{ENV.cflags}" if OS.mac?)
```
Comment (sjackman) [inline Formula/man-db.rb]: What's the reason for this change? What's the error message without it?

File: Formula/man-db.rb (hunk 1 above)
Comment (maxim-belkin) [reply Formula/man-db.rb]: https://gist.github.com/0f4ff214c6e148957a7211a36307a6f7

> relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/home/linuxbrew/.linuxbrew/bin/ld: final link failed: nonrepresentable section on output
collect2: error: ld returned 1 exit status
Makefile:1586: recipe for target 'libman.la' failed

Comment (iMichka) [reply Formula/man-db.rb]: This error message is often triggered when a .so file is missing somewhere.

Is there a reason why we use:
`--enable-static --disable-shared` instead of `--enable-static --enable-shared`?

File: Formula/man-db.rb (hunk 2 above)
Comment (maxim-belkin) [reply Formula/man-db.rb]: https://gist.github.com/75c0376ae85817c69e07ee90f7647a67

> /home/linuxbrew/.linuxbrew/bin/ld: -f may not be used without -shared
collect2: error: ld returned 1 exit status
Makefile:1860: recipe for target 'catman' failed
make[3]: *** [catman] Error 1

**Comment by maxim-belkin** [comment]
From CircleCI:
> failed: brew test man-db --verbose

        Testing man-db
==> /home/linuxbrew/.linuxbrew/Cellar/man-db/2.8.5_1/bin/gman true
/home/linuxbrew/.linuxbrew/Cellar/man-db/2.8.5_1/bin/gman: can't execute tbl: No such file or directory
/home/linuxbrew/.linuxbrew/Cellar/man-db/2.8.5_1/bin/gman: can't execute nroff: No such file or directory

I'll make `groff` full dependency

File: Formula/man-db.rb (hunk 1 above)
Comment (maxim-belkin) [reply Formula/man-db.rb]: https://github.com/Homebrew/homebrew-core/pull/36629/files#r253271581

**Comment by maxim-belkin** [comment]
rebased on top of master

File: Formula/man-db.rb (hunk 2 above)
Comment (sjackman) [reply Formula/man-db.rb]: This error appears to be caused by adding `-fPIC` to the command line above with `ENV.append_to_cflags "-fPIC"`. I'd suggest using either `with_env` or `remove_from_cflags` to remove `-fPIC` after building `libpipeline`.

Comment (sjackman) [reply Formula/man-db.rb]: Alternatively we could add `--enable-static --disable-shared` to `man-db` as well as `libpipeline` and get rid of `-fPIC`. The downside is we would no longer build the shared libraries `libman-2.8.4.so  libmandb-2.8.4.so`. Does anything use those libraries though? My guess is not.

Comment (sjackman) [reply Formula/man-db.rb]: I'm personally fine with either solution.

Comment (maxim-belkin) [reply Formula/man-db.rb]: The `f` it complains about is not from `-fPIC` but from `-flat_namespace` which is not known to `ld` on Linux.

Comment (maxim-belkin) [reply Formula/man-db.rb]: Here are gist-logs where I keep `-fPIC` but remove `-flat_namespace`:
https://gist.github.com/d93f0f179b60e03ca23de84ce004f514

Comment (maxim-belkin) [reply Formula/man-db.rb]: https://gist.github.com/maxim-belkin/d93f0f179b60e03ca23de84ce004f514#file-05-make-L4
This is `make CFLAGS` line that shows that `-fPIC` is present and the build succeeds.

File: Formula/man-db.rb [hunk 3]
```diff
@@ -95,7 +101,12 @@ def caveats; <<~EOS
...
     output = shell_output("#{bin}/gman true")
-    assert_match "BSD General Commands Manual", output
-    assert_match "The true utility always returns with exit code zero", output
+    if OS.mac?
+      assert_match "BSD General Commands Manual", output
+      assert_match "The true utility always returns with exit code zero", output
+    else
+      assert_match "true - do nothing, successfully", output
+      assert_match "GNU coreutils online help: <http://www.gnu.org/software/coreutils/", output
+    end
```
Comment (maxim-belkin) [inline Formula/man-db.rb]: This is good for now but I think we ought to come up with a different test and submit it upstream.

File: Formula/man-db.rb (hunk 2 above)
Comment (sjackman) [reply Formula/man-db.rb]: In that case, this line should be fixed:
```ruby
    # NB: Remove once man-db 2.8.6 is released
    # https://git.savannah.gnu.org/cgit/man-db.git/commit/?id=056e8c7c012b00261133259d6438ff8303a8c36c
    ENV.append_to_cflags "-Wl,-flat_namespace,-undefined,suppress" if OS.mac?
```

Comment (maxim-belkin) [reply Formula/man-db.rb]: 👍  will do that.

Shall I clean up CFLAGS after installing libpipeline (it does _not_ affect the build)?

In principle, we could:
1. create `ENV.fix_cflags` to remove flags unsupported by `ld`,
2. instruct `ENV.append_to_cflags` to ignore unsupported flags (I'm not a fan of this option, but...)

Comment (sjackman) [reply Formula/man-db.rb]: > Shall I clean up CFLAGS after installing libpipeline (it does not affect the build)?

I believe it does affect the build, because it's compiling the executables using `-fPIC`, which in principle results in slower and larger executables, but realistically probably doesn't make a difference. I think it is cleaner to clean up `CFLAGS` and remove `-fPIC`.

Currently it's Superenv's job to remove unsupported flags.
See https://github.com/Homebrew/brew/blob/9a1936cd4b773dbd74695d3c110583570b58771b/Library/Homebrew/shims/super/cc#L179

File: Formula/man-db.rb [hunk 4]
```diff
@@ -12,6 +12,11 @@ class ManDb < Formula
   end
 
   depends_on "pkg-config" => :build
+  unless OS.mac?
+    depends_on "groff"
+    depends_on "gdbm"
+    depends_on "zlib"
+  end
```
Comment (sjackman) [inline Formula/man-db.rb]: These can be changed to `uses_from_macos "…"`.

Comment (maxim-belkin) [reply Formula/man-db.rb]: Didn't @issyl0 had an issue with `uses_from_macos` in one of her PRs recently?

Comment (maxim-belkin) [reply Formula/man-db.rb]: Switched to `uses_from_macos`

**Comment by maxim-belkin** [comment]
Thanks, Shaun.
I don't think I can push to Homebrew, so let me ping @iMichka.

@iMichka, ping! :)

**Comment by maxim-belkin** [comment]
🎉

**Comment by iMichka** [comment]
Thanks @maxim-belkin !
